### ISSUE TITLE
Dice over all samples

### ISSUE THREAD

**Issue by shpavisaslan** [jocicmarko/ultrasound-nerve-segmentation#67]
I am new in Keras. I am trying to use your code in brain tumor segmentation but I confused about the dice metric. I don't know actually this code measures the dice for each image and then measures the mean value over all samples or it just measures the dice over all sample once. What I mean, consider that I have an input like (3000,218,218) includes 3000 samples. is it calculate the dice for each 218*218 image and then measure the mean over all 3000 samples or it convert this huge matrix to a vector and the measure the dice for all samples once. 
Thanks

**Comment by jocicmarko** [comment]
Hi, welcome to Keras community :).
The code I provided calculates Dice metric over whole batch - it doesn't average over samples.

**Comment by shpavisaslan** [comment]
Thanks for your response. I understand now. So, in this way, when I give a validation set, it measure the dice again over set! 
I have another question! I wrote the your code using np insted of keras values. However, the dice value is different. What do you think! is there any advantage to use keras variables than numpy!?

**Comment by jocicmarko** [comment]
> So, in this way, when I give a validation set, it measure the dice again over set!

Exactly, so this is not a reliable metric for reporting (in a paper for example). I suggest that for validation/test you calculate the Dice metric per image (batch size 1) and then average over those.

> I wrote the your code using np insted of keras values. However, the dice value is different. What do you think! is there any advantage to use keras variables than numpy!?

How large is the difference? It's normal that there is some difference, usually because default numpy uses float64 precision and Tensorflow uses float32 precision. However, if the difference is large, it might be that the implementations are different.

**Comment by shpavisaslan** [comment]
For instance; for 1 epoch in Keras I have dice over whole images in validation set like 0.0442, but when I save the model exactly related to that epoch and try to use model.predict function over whole same validation set, I have dice value (Numpy) around 0.0698!!
The difference is high actually if you look some other metrics like precision!! 
Keras (precision:0.3563)
Numpy(recall: 0.9464)

def precision(y_true, y_pred):
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
    precision = true_positives / (predicted_positives + K.epsilon())
    return precision

def precision(y_true, y_pred):
   true_positives = np.sum(np.round(np.clip(flair_imgs_mask_validation * result, 0, 1)))
   predicted_positives = np.sum(np.round(np.clip(result, 0, 1)))
   precision = true_positives / (predicted_positives)
    return precision

**Comment by shpavisaslan** [comment]
I found some bugs in my code. So basically the values are more or less same with a few differences that can be ignored I think. 
dice (Keras): 0.002789
dice (Numpy): 0.002946
and the other metrics is more or less like this in initial epochs.
